flang/lib/Optimizer/Dialect/FIROps.cpp

```diff
@@ -37,6 +37,15 @@ namespace {
 #include "flang/Optimizer/Dialect/CanonicalizationPatterns.inc"
 } // namespace
 
+static void addVolatileMemoryEffects(
+    mlir::Type type, llvm::SmallVectorImpl<mlir::SideEffects::EffectInstance<
+                         mlir::MemoryEffects::Effect>> &effects) {
+  if (fir::isa_volatile_ref_type(type)) {
+    effects.emplace_back(mlir::MemoryEffects::Read::get());
+    effects.emplace_back(mlir::MemoryEffects::Write::get());
```

Comment:
What was your reasoning for adding both read and write? Maybe this should be modeled as unknown memory effects?

Reply:
There is a precedent with  the LLVM dialect: https://github.com/llvm/llvm-project/commit/feb7beaf70bace1c3ffafc7f732c8fadca5e8c9d

It cites LLVM documentation that mentions Volatile read are considered to can have any read/write effects:
See https://llvm.org/docs/LangRef.html#volatile-memory-accesses

Reply:
It may be okay to follow the precedent in LLVM dialect, but I think there is a room for improvement.  Per LLVM LangRef, `volatile` does allow reordering non-volatile operations with regards to volatile operations, so I wonder if volatile operations should be reading/writing from/to a dedicated `volatile` (or `inaccessible`) resource.

No need to address it in this PR.

Reply:
Thanks for the answers everyone.

I meant "unknown memory effects" as in not specifying any at all. This would force passes to be as pessimistic as possible about this operation as they would have no way to reason about it.

However, I now think Slava is right that being maximally pessimistic could be overkill in the long run. Anyway, doing the same as the LLVM dialect is great for this PR.

Reply:
Adding a memory resource for volatile operations was easier than I expected. It will be in the next shared revision. Thanks for the idea Slava!